### PR TITLE
TransportVersionUtils#randomVersionBetween does not work with version extensions (106119)

### LINKED ISSUE #106119: TransportVersionUtils#randomVersionBetween does not work with version extensions
When a version constant is provided by a version extension, `TransportVersionUtils#randomVersionBetween` will fail with a message similar to the following
```
java.lang.IllegalArgumentException: maxVersion [8599010] does not exist.
```
Such a version constant can be automatically returned by calling `TransportVersion.current()` if a version extension happens to define the highest version.

### PR DESCRIPTION
**Pull request by alexey-ivanov-es**
Introduces a new extension method to VersionExtension enabling extensions to provide additional versions and creates method TransportVersion.getAllVersions returning all transport versions defined by Elasticsearch and the extension. This ensures that TransportVersion.current() always returns the correct current (latest) transport version even if it is defined by an extension.

Fixes #106119

### CODE REVIEW

**Review by ldematte — COMMENTED** [review]
Technically this fixes the bug, but the purpose of the function is to test (randomly) any possible version in a range. For serverless, that means more versions (potentially) Shouldn't we add the versions from the extensions instead of just adjusting the index?

**Comment by alexey-ivanov-es** [comment]
@ldematte The list of versions is currently only available in a static context. Here are the options I see for making it extendable:

1. Create a service to manage all known versions, which would involve a significant refactoring
2. Take a similar approach to org.elasticsearch.TransportVersion#current - introduce a new method in org.elasticsearch.internal.VersionExtension to return all versions, call it from static context and merge all versions into a single list

I feel that existing approach to version extensions is error-prone, but as we only need the list of all versions for unit tests now, I’m not sure it’s worth the effort to fix it properly. What do you think? Am I missing something?

**Comment by ldematte** [comment]
I was thinking about 2: using the current extension mechanism, extending it if needed.
You are right in saying that we have to decide if it's worth it or not. Probably ... no. Which means I'm fine with the current code. 
But I feel we could ask for a second opinion. @thecoop, wdyt?

**Comment by thecoop** [comment]
Could we have a `ServerlessTransportVersionUtils` that explicitly knows about serverless things? Then in the main ES one we can throw if the current version is not a main version, so serverless uses of that class are pointed towards the serverless-specific class.

There's also a question about which versions are actually 'live', which requires knowing about deployed serverless versions. Maybe we could use a hard limit - go back a maximum of 50 versions?

**Comment by ldematte** [comment]
++ for `ServerlessTransportVersionUtils`

**Comment by ldematte** [comment]
There is still the problem of unchanged tests running on the serverless codebase; even when we introduce ServerlessTransportVersionUtils, I think the change in this PR would be still needed.

**Comment by alexey-ivanov-es** [comment]
> Could we have a ServerlessTransportVersionUtils that explicitly knows about serverless things? 
Does this solve the original problem? It seems to me that if core tests using TransportVersionUtils#randomVersionBetween are executed as part of the serverless build, then TransportVersion#current might still return a version that doesn’t exist in core

>There's also a question about which versions are actually 'live', which requires knowing about deployed serverless versions. Maybe we could use a hard limit - go back a maximum of 50 versions?
I didn’t quite understand this part. Could you clarify where this question comes up and in what context we would need to go back?

**Comment by thecoop** [comment]
The issue is that `TransportVersionUtils.versionsBetween` statically references `ALL_VERSIONS`, which does not include serverless. So there are two options:

1. Include the set of all versions in the data injected by `VersionExtension`, and use that.
2. Have a separate `ServerlessTransportVersionUtils` class that statically knows about serverless transport versions, and blow up if this `TransportVersionUtils` is used from serverless.

1 is perhaps the more dev-friendly option, and adding another method to `VersionExtension` is not a bad thing.
2 doesn't need any cross-repo changes between the two.

Additionally, we can think more about what transport versions is likely to be used in serverless whilst doing this change. Because we're always releasing off main, we don't need to test 8.1, 8.2, etc in the serverless repo. Using methods 1 or 2, we can limit the transport versions that are returned by any methods in `TransportVersionUtils` or `ServerlessTransportVersionUtils` to those that are likely to be hit in prod serverless - which are only the versions added in the last ~4 weeks.

**Comment by alexey-ivanov-es** [comment]
@elasticsearchmachine update branch

**Comment by alexey-ivanov-es** [comment]
@elasticsearchmachine update branch

**Comment by alexey-ivanov-es** [comment]
@elasticmachine update branch

**Comment by alexey-ivanov-es** [comment]
I decided that it is better to implement this properly with an extension method.
1. When tests written in elasticsearch repository (and therefore working with TransportVerionUtils) are executed as part of serverless build, they still might have serverless version as current (==max) and we still need protection in TransportVerionUtils for such cases. Extension solves this.
2. ServerlessTransportVersionUtils actually, wouldn't solve any problem at the moment, because currently there are no tests in the serverless repository that use TransportVersionUtils, so there is just no place to use it

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/TransportVersions.java:346]: This is no longer a builder-type object now

File: server/src/main/java/org/elasticsearch/internal/VersionExtension.java
Comment (alexey-ivanov-es) [reply server/src/main/java/org/elasticsearch/internal/VersionExtension.java]: I think `getAllTransportVersions()` might be slightly misleading, because the method doesn't return all known transport versions, it returns transport versions defined by extension, so I would prefer to keep name as `getTransportVersions()`.

I removed a default implementation

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (alexey-ivanov-es) [reply server/src/main/java/org/elasticsearch/TransportVersions.java]: Renamed

**Comment by alexey-ivanov-es** [comment]
@elasticmachine update branch

**Review by thecoop — APPROVED** [review]
LGTM, if @rjernst is also ok with it.

This does mean that `TransportVersions.VERSION_IDS` will contain constants not defined in `TransportVersions`, but I think thats ok given the situations that will happen in.

**Comment by rjernst** [comment]
> This does mean that TransportVersions.VERSION_IDS will contain constants not defined in TransportVersions

I don't think we should do that. `TransportVersions` are supposed to be the versions defined in server, and `ServerlessTransportVersions` are those defined in serverless. We should keep them separate.

Can we move the "getl all transport versions" on to `TransportVersion`? That class is supposed to be what handles talking with the extension, which is why the current holder is there (and I think it should remain there).

**Comment by alexey-ivanov-es** [comment]
>I don't think we should do that. TransportVersions are supposed to be the versions defined in server, and ServerlessTransportVersions are those defined in serverless. We should keep them separate.

What is the purpose of having a list of transport versions in a serverless app which doesn't include serverless versions?

>Can we move the "getl all transport versions" on to TransportVersion?

I think yes. To me, `TransportVersion` doesn’t sound like the right place for a method `getAllVersions()`, but I’m not strongly opposed to it either

**Comment by rjernst** [comment]
There's some history here. TransportVersion is the "thing" that we want. It has the `current()` method, and used to have an "all" type method. The constants all used to exist on TransportVersion.

We created TransportVersions to avoid static initialization problems. The constants are defined separately from the type of the constants. ServerlessTransportVersions is then the serverless equivalent, where those constants are defined.

The reason I think the "all" type method belongs on TransportVersion is because that is where `current()` lives. It is the pivot point that knows how to (in a way that won't break class init) get all the version constants. TransportVersions and ServlerlessTransportVersions should be simple containers of constants. That's not exactly the case right now, but I think we should move back toward that direction.

**Comment by alexey-ivanov-es** [comment]
Moved getAllVersions to TransportVersion

**Review by rjernst — COMMENTED** [review]
This looks pretty good. I have just a couple more thoughts.

File: test/framework/src/main/java/org/elasticsearch/KnownTransportVersions.java
Comment (thecoop) [reply test/framework/src/main/java/org/elasticsearch/KnownTransportVersions.java:19]: Or it can be a `SequencedSet`?

Comment (alexey-ivanov-es) [reply test/framework/src/main/java/org/elasticsearch/KnownTransportVersions.java:19]: Do we need this class at all? Can we use `TransportVersion.getAllVersions()` instead?

Comment (alexey-ivanov-es) [reply test/framework/src/main/java/org/elasticsearch/KnownTransportVersions.java:19]: Regarding `SequencedSet` - the result of this method is typically used as `List`: we perform operations like subList, binarySearch, and accessing elements by index. So, I think it’s better to keep it as `List`

Comment (rjernst) [reply test/framework/src/main/java/org/elasticsearch/KnownTransportVersions.java:19]: We can probably get rid of KnownTransportVersions

**Comment by alexey-ivanov-es** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/TransportVersions.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/TransportVersions.java:263]: nit: unmodifiableList creates a wrapper which makes every access indirect. Instead, use `List.copyOf` to get an unmodifiable list from another collection.

File: server/src/main/java/org/elasticsearch/TransportVersion.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/TransportVersion.java:179]: Same here, use `List.copyOf(allVersions)`

File: server/src/main/java/org/elasticsearch/TransportVersion.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/TransportVersion.java:182]: nit: toMap -> toUnmodifiableMap

**Review by rjernst — APPROVED** [review]
A couple last nits, but otherwise looks good.

**Review by thecoop — APPROVED** [review]

**Pull request merged**
